### ISSUE TITLE
Compiler arguments are not properly ignored during checks

### ISSUE THREAD

**Issue by ePirat** [mesonbuild/meson#9866]
**Describe the bug**
As stated in the [meson documentation for the compile object](https://mesonbuild.com/Reference-manual_returned_compiler.html#compiler-object-compiler), for functions that do checks:

> These compiler checks do not use compiler arguments added with add_*_arguments(), via -Dlang_args on the command-line, or through CFLAGS/LDFLAGS, etc in the environment. Hence, you can trust that the tests will be fully self-contained, and won't fail because of custom flags added by other parts of the build file or by users.

However this is not actually the case, meson only ignores arguments set with `add_*_arguments()`, not for the other cases.

**To Reproduce**
See #9865 

**Expected behavior**
The compiler check functions should either behave as documented or the documentation should be adjusted.

I am not sure whats the best route to take here to fix this.

**Comment by jpakkane** [comment]
FWICT the behaviour is correct and the docs are wrong. Since lang args (as well as CFLAGS, which go to lang args) are used for all compilations, they should also be used for the compilation tests.

**Comment by ePirat** [comment]
Yeah I think that makes sense, as for example when cross-compiling to iOS you usually have:

```ini
[built-in options]
c_args = ['-arch', 'arm64', '-miphoneos-version-min=11.0', '-isysroot', '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk']
cpp_args = ['-arch', 'arm64', '-miphoneos-version-min=11.0', '-isysroot', '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk']
c_link_args = ['-arch', 'arm64', '-miphoneos-version-min=11.0', '-isysroot', '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk']
cpp_link_args = ['-arch', 'arm64', '-miphoneos-version-min=11.0', '-isysroot', '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk']
objc_args = ['-arch', 'arm64', '-miphoneos-version-min=11.0', '-isysroot', '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk']
objcpp_args = ['-arch', 'arm64', '-miphoneos-version-min=11.0', '-isysroot', '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk']
```

Not honoring these during checks would break things.

**Comment by keith-packard** [comment]
I've moved settings required for the compiler to work properly to the [binaries] part of the machine/cross file -- those aren't included in the list of things ignored by compiler checks, and also allow the user to specify other options (like warnings or optimization settings) via the -Dc_args command line option without that overriding entries in the [properties] section.

**Comment by amyspark** [comment]
We at Krita ran into this during our Meson integration work: https://invent.kde.org/graphics/krita/-/merge_requests/1607#note_595460

Used the https://github.com/picolibc/picolibc/pull/263 approach, we'll see if this works.
